eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs

```diff
@@ -1771,3 +1803,60 @@ logConwayTxBalance tx = do
   certState <- getsNES $ nesEsL . esLStateL . lsCertStateL
   utxo <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
   logString $ showConwayTxBalance pp certState utxo tx
+
+submitBootstrapAwareFailingVote ::
+  ConwayEraImp era =>
+  Vote ->
+  Voter (EraCrypto era) ->
+  GovActionId (EraCrypto era) ->
+  SubmitFailureExpectation era ->
+  ImpTestM era ()
+submitBootstrapAwareFailingVote vote voter gaId =
+  submitBootstrapAwareInternal
+    (submitVote_ vote voter gaId)
+    (submitFailingVote voter gaId)
+
+submitBootstrapAwareFailingProposal ::
+  ConwayEraImp era =>
+  ProposalProcedure era ->
+  SubmitFailureExpectation era ->
+  ImpTestM era (StrictMaybe (GovActionId (EraCrypto era)))
+submitBootstrapAwareFailingProposal proposal =
+  submitBootstrapAwareInternal
+    (SJust <$> submitProposal proposal)
+    (submitFailingProposal proposal >=> const (pure SNothing))
+
+data SubmitFailureExpectation era
+  = FailBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
+  | FailPostBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
+  | FailBootstrapAndPostBootstrap
+      (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
+      (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
+
+pattern FailBoth ::
+  NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)) ->
+  NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)) ->
+  SubmitFailureExpectation era
+pattern FailBoth {bootstrapFailures, postBootstrapFailures} =
+  FailBootstrapAndPostBootstrap bootstrapFailures postBootstrapFailures
```

Comment:
What's the point of this type synonym? It is identical to `FailBootstrapAndPostBootstrap`

Reply:
In order to be able to give names  for `bootstrapFailures` and `postBootstrapFailures`. 
If i do this:
```
data SubmitFailureExpectation era
  = FailBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailPostBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailBootstrapAndPostBootstrap
      { bootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
      , postBootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
     } 
```
I get a compiler warning : "    Use of partial record field selector: ‘bootstrapFailures’" 

I guess without names it's pretty clear which is which, but I thought since they have the same type, it would be better to be able to pass with names the values.

Reply:
Using record syntax for sum types is indeed not a good idea, hence the compiler warning. With pattern synonym you have the same issue, you just manage to shut the compiler up.

The correct way would be to create an extra type (less overhead that pattern synonym too):

```
data SubmitFailureExpectation era
  = FailBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailPostBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailBootstrapAndPostBootstrap (FailBoth era)

data FailBoth era = FailBoth
      { bootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
      , postBootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
     } 
```

Reply:
Interesting!  Thank you